distribution/pull_v2_windows.go

```diff
@@ -133,21 +132,43 @@ func (mbv manifestsByVersion) Swap(i, j int) {
 	mbv.list[i], mbv.list[j] = mbv.list[j], mbv.list[i]
 }
 
-// checkImageCompatibility blocks pulling incompatible images based on a later OS build
+// checkImageCompatibility tries to check whether a Windows image is incompatible
+// with the host's OS version. It is a no-op for non-Windows images. It returns
+// an error if the image is incompatible, but ignores failures when parsing
+// imageOSVersion.
+//
 // Fixes https://github.com/moby/moby/issues/36184.
 func checkImageCompatibility(imageOS, imageOSVersion string) error {
```

Comment:
This function doesn't have/receive enough information to actually make this comparison accurately -- as seen on https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility, the compatibility of a given image depends not only on the host/container versions, but _also_ process vs hyper-v isolation. :sob:

So, which column of that "compatibility" matrix is this actually checking?

Reply:
Oh! Good point. Hm...  😭

Reply:
The hcsshim-side function being used here is for process isolation. Hyper-V isolation is _expected_ to be a simple "less-than-or-equal-to" check; from time-to-time people have harangued it into running _newer_ containers on older host, but it's not in any way promised to work correctly.

Hyper-V isolation is a micro-VM like WSL2, running its own OS kernel, so the compatibility layer there is simply in the VM/host communication, i.e. newer HCS knows how to set up VMs/network etc for older OS containers.